Stratis.Bitcoin/BlockPulling/BlockPuller.cs

```diff
@@ -319,36 +208,46 @@ private void AssignPendingVectors()
             if (innernodes.Length == 0)
                 return;
 
-            List<InventoryVector> inventoryVectors = new List<InventoryVector>();
-            uint256 result;
-            while (this.pendingInventoryVectors.TryTake(out result))
-            {
-                inventoryVectors.Add(new InventoryVector(InventoryType.MSG_BLOCK, result));
-            }
+            if (this.pendingInventoryVectors.Count == 0)
+                return;
 
+            int minHeight = int.MaxValue;
             Dictionary<int, InventoryVector> vectors = new Dictionary<int, InventoryVector>();
-            var minHeight = int.MaxValue;
-            foreach (InventoryVector vector in inventoryVectors.ToArray())
+
+            uint256 blockHash;
+            while (true)
```

Comment:
Hmm this is interesting, you can avoid the loop on a lock in this case, as the result will be almost the same if another threads comes in.  

In this code two (or more) threads can deplete the queue. and call the distributor twice.
 
You can pull the lock out of the loop and populate a local list to deplete the queue , if another thread will come it only waits for the deplete code and returns (as the queue will be empty)

Then you don't have to lock on every iteration.